File: python/relink_py/examples/ros2_compare/relink_camera_sub_raw.py
#!/usr/bin/env python3
"""ReLink camera subscriber counterpart to relink_camera_pub_raw.py --
displays the live feed in a cv2 window and prints running latency/FPS/
bandwidth stats, in the same format as ros2_camera_sub.py, so the two
printed stat streams can be compared directly. Auto-detects raw vs
JPEG-compressed payloads (a JPEG frame decodes; a raw frame doesn't and
falls back to reshaping by --width/--height), so the same subscriber
works against either publisher mode.

Requires OpenCV: pip install opencv-python

Run:
    python3 relink_camera_sub_raw.py [rlcore_ip] [--width W] [--height H] [--compressed]
"""
import sys
import os
import argparse
import struct
import time

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", ".."))
from relink import RelinkNode

try:
    import cv2
    import numpy as np
except ImportError:
    print("relink_camera_sub_raw.py requires OpenCV -- pip install opencv-python", file=sys.stderr)
    sys.exit(1)

TOPIC_CAMERA = "/compare/camera"
HEADER_LEN = 8  # struct 'd'


def parse_args():
    p = argparse.ArgumentParser()
    p.add_argument("rlcore_ip", nargs="?", default=None)
    p.add_argument("--width", type=int, default=320)
    p.add_argument("--height", type=int, default=240)
    p.add_argument("--compressed", action="store_true")
    return p.parse_args()


def main():
    args = parse_args()

    node = RelinkNode()
    if args.rlcore_ip:
        node.set_rlcore.ip(args.rlcore_ip)
    else:
        node.use_multicast_discovery()

    recv_bytes = 0
    recv_count = 0
    latencies = []
    start = time.time()

    def on_frame(frame_id, data):
        nonlocal recv_bytes, recv_count
        send_time, = struct.unpack("<d", data[:HEADER_LEN])
        latency_ms = (time.time() - send_time) * 1000.0
        latencies.append(latency_ms)

        body = data[HEADER_LEN:]
        if args.compressed:
            arr = cv2.imdecode(np.frombuffer(body, dtype=np.uint8), cv2.IMREAD_COLOR)
        else:
            arr = np.frombuffer(body, dtype=np.uint8).reshape((args.height, args.width, 3))
        cv2.imshow("ReLink camera", arr)
        cv2.waitKey(1)

        recv_bytes += len(data)
        recv_count += 1
        if recv_count % 30 == 0:
            elapsed = time.time() - start
            avg_lat = sum(latencies[-30:]) / len(latencies[-30:])
            print(f"relink: {recv_count} frames, {recv_count/elapsed:.1f} fps, "
                  f"{recv_bytes/elapsed/1e6:.2f} MB/s, avg latency {avg_lat:.2f} ms, "
                  f"{len(data)} bytes/frame")

    node.subscribe_image(TOPIC_CAMERA, on_frame)
    print(f"relink_camera_sub_raw: subscribed to {TOPIC_CAMERA}, displaying live "
          "(ctrl-c to quit)")
    node.spin()


if __name__ == "__main__":
    sys.exit(main() or 0)

File: cpp/examples/ros2_compare/relink_camera_sub_raw.cpp
// ReLink camera subscriber, C++ counterpart to relink_camera_pub_raw.cpp
// and python/relink_py/examples/ros2_compare/relink_camera_sub_raw.py.
// Displays the live feed in an OpenCV window and prints running
// latency/FPS/bandwidth stats in the same format as the ROS2 side, for
// direct comparison. Auto-decodes JPEG (--compressed), GPU H264
// (--h264, must match the publisher), or reshapes raw BGR bytes by
// --width/--height, matching whichever mode the publisher was run with.
//
// Requires OpenCV. Build:
//   g++ -std=c++17 -O2 -I ../../relink/include -pthread relink_camera_sub_raw.cpp -o relink_camera_sub_raw $(pkg-config --cflags --libs opencv4)
// Run:
//   ./relink_camera_sub_raw [rlcore_ip] [--width W] [--height H] [--compressed] [--h264]

#include "relink/relink.hpp"
#include "gst_h264_codec.hpp"
#include <opencv2/opencv.hpp>
#include <cstdio>
#include <cstring>
#include <cstdint>
#include <memory>
#include <vector>
#include <string>
#include <chrono>
#include <deque>
#include <mutex>
#include <numeric>

using namespace std::chrono;

static const char* TOPIC_CAMERA = "/compare/camera";

struct Args {
    std::string rlcore_ip;
    int width = 320, height = 240;
    bool compressed = false;
    bool h264 = false;
};

static Args parse_args(int argc, char** argv) {
    Args a;
    std::vector<std::string> pos;
    for (int i = 1; i < argc; ++i) {
        std::string arg = argv[i];
        if (arg == "--width") a.width = std::atoi(argv[++i]);
        else if (arg == "--height") a.height = std::atoi(argv[++i]);
        else if (arg == "--compressed") a.compressed = true;
        else if (arg == "--h264") a.h264 = true;
        else pos.push_back(arg);
    }
    if (!pos.empty()) a.rlcore_ip = pos[0];
    return a;
}

static double now_sec() {
    return duration_cast<duration<double>>(system_clock::now().time_since_epoch()).count();
}

int main(int argc, char** argv) {
    Args args = parse_args(argc, argv);

    RelinkNode node;
    if (!args.rlcore_ip.empty()) node.set_rlcore.ip(args.rlcore_ip);
    else node.use_multicast_discovery();

    uint64_t recv_bytes = 0;
    uint32_t recv_count = 0;
    std::deque<double> latencies;
    std::mutex stats_mutex;
    auto start = steady_clock::now();

    // --h264: decode happens on a background thread (see
    // gst_h264_codec.hpp) -- imshow is called from that thread's
    // callback since it's the only thread that ever touches the
    // decoded-frame window.
    std::unique_ptr<gst_h264::Decoder> decoder;
    if (args.h264) {
        decoder = std::make_unique<gst_h264::Decoder>(
            args.width, args.height,
            [w = args.width, h = args.height](const uint8_t* data, size_t len) {
                (void)len;
                cv::Mat frame(h, w, CV_8UC3, const_cast<uint8_t*>(data));
                cv::imshow("ReLink camera", frame);
                cv::waitKey(1);
            });
    }

    node.subscribe_image(TOPIC_CAMERA, [&](uint32_t /*id*/, const std::vector<uint8_t>& data) {
        double send_time;
        std::memcpy(&send_time, data.data(), 8);
        double latency_ms = (now_sec() - send_time) * 1000.0;

        const uint8_t* body = data.data() + 8;
        size_t body_len = data.size() - 8;

        if (args.h264) {
            decoder->push_nal(body, body_len);
        } else {
            cv::Mat frame;
            if (args.compressed) {
                std::vector<uint8_t> jpeg(body, body + body_len);
                frame = cv::imdecode(jpeg, cv::IMREAD_COLOR);
            } else {
                frame = cv::Mat(args.height, args.width, CV_8UC3, const_cast<uint8_t*>(body)).clone();
            }
            cv::imshow("ReLink camera", frame);
            cv::waitKey(1);
        }

        std::lock_guard<std::mutex> lock(stats_mutex);
        latencies.push_back(latency_ms);
        if (latencies.size() > 30) latencies.pop_front();
        recv_bytes += data.size();
        ++recv_count;
        if (recv_count % 30 == 0) {
            double elapsed = duration<double>(steady_clock::now() - start).count();
            double avg_lat = std::accumulate(latencies.begin(), latencies.end(), 0.0) / latencies.size();
            std::printf("relink: %u frames, %.1f fps, %.2f MB/s, avg latency %.2f ms, %zu bytes/frame\n",
                        recv_count, recv_count / elapsed, recv_bytes / elapsed / 1e6, avg_lat, data.size());
        }
    });

    std::printf("relink_camera_sub_raw: subscribed to %s, displaying live (ctrl-c to quit)\n", TOPIC_CAMERA);
    node.spin();
}
